src/vm/database/structures.rs

```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
+    pub amount_unlocked: u128,
+    pub amount_locked: u128,
+    pub unlock_height: u64,
+}
+clarity_serializable!(STXBalance);
+
+#[derive(Debug)]
+pub enum STXBalanceError {
+    Overflow,
+    Underflow,
+    LockActive,
+}
+
+type Result<T> = std::result::Result<T, STXBalanceError>;
+
+impl STXBalance {
+
+    pub fn zero() -> STXBalance {
+        STXBalance {
+            amount_unlocked: 0,
+            amount_locked: 0,
+            unlock_height: 0,
+        }
+    }
+
+    pub fn get_available_balance_at_block(&self, block_height: u64) -> u128 {
+        match self.has_locked_tokens_unlockable(block_height) {
+            true => self.get_total_balance(),
+            false => self.amount_unlocked
+        }
+    }
+
+    pub fn lock_tokens(&mut self, amount_to_lock: u128, unlock_height: u64, current_height: u64) -> Result<()> {
+        let unlocked = self.unlock_available_tokens_if_any(current_height);
+        if  unlocked > 0 {
+            debug!("Consolidated after account-token-lock");
+        }
+
+        // todo(ludo): Also check that unlock_height > current_height
+        if self.has_locked_tokens(current_height) {
+            return Err(STXBalanceError::LockActive)
+        }
+
+        self.unlock_height = unlock_height;
+        self.amount_unlocked = self.amount_unlocked.checked_sub(amount_to_lock)
+            .expect("STX overflow");
+        self.amount_locked = amount_to_lock;
+        Ok(())
+    }
+
+    pub fn unlock_available_tokens_if_any(&mut self, block_height: u64) -> u128 {
+        if !self.has_locked_tokens_unlockable(block_height) {
+            return 0
+        }
+
+        let unlocked = self.amount_locked;
+        self.unlock_height = 0;
+        self.amount_unlocked = self.amount_unlocked.checked_add(unlocked)
+            .expect("STX overflow");
+        self.amount_locked = 0;
+        unlocked
+    }
+
+    pub fn get_total_balance(&self) -> u128 {
+        self.amount_unlocked.checked_add(self.amount_locked)
+            .expect("STX overflow")
+    }
+
+    pub fn has_locked_tokens(&self, block_height: u64) -> bool {
+        self.amount_locked > 0 && self.unlock_height > block_height
+    }
+
+    pub fn has_locked_tokens_unlockable(&self, block_height: u64) -> bool {
+        self.amount_locked > 0 && self.unlock_height <= block_height
+    }
+
+    pub fn can_transfer(&self, amount: u128, block_height: u64) -> bool {
+        self.get_available_balance_at_block(block_height) >= amount
+    }
+
+    pub fn debit(&mut self, amount: u128, block_height: u64) -> Result<()> {
+        let unlocked = self.unlock_available_tokens_if_any(block_height);
+        if  unlocked > 0 {
+            debug!("Consolidated after account-debit");
+        }
+
+        self.amount_unlocked = self.amount_unlocked.checked_sub(amount)
+            .ok_or_else(|| STXBalanceError::Underflow)?;
+        Ok(())
+    }
+
+    pub fn credit(&mut self, amount: u128, block_height: u64) -> Result<()> {
```

Comment:
Why does `credit` take a block height?  Also, why does crediting tokens also unlock them?  That seems unfair to the token sender -- it forces the token sender to pay to unlock the recipient's tokens.

Reply:
Yes, because previously, unlocking tokens for the recipient would have a cost (you'd have to update the MARFwith the new `locked_amount `, `unlock_height`). 
Now that `(unlocked_amount, locked_amount, unlock_height)` are stored under the same MARF key, we can use the `unlock_amount` update as a vehicle for also updating the values `locked_amount `, `unlock_height`.

Reply:
>  unlocking tokens for the recipient would have a cost

No, it wouldn't.  The _sender_ would pay to unlock its tokens before sending them to the recipient.  In the previous version, the recipient's unlocked balance would increment, but the lock state for the recipient wouldn't change.

Reply:
Yes that's what I'm saying

> it forces the token sender to pay to unlock the recipient's tokens. I intentionally did not do this originally

With the previous approach, materializing the unlock in the MARF for the recipient would cost to the sender, and this would have been unfair for the sender. 
After addressing the change request https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r479542332, updating the lock for the recipient is no longer an extra cost for the sender.

Reply:
> With the previous approach, materializing the unlock in the MARF for the recipient would cost to the sender, and this would have been unfair for the sender.

The sender paid only to unlock its own tokens in the original PR.  That was intentional.  The recipient's lock state was not touched.

Here is the reproduced `stx_transfer_consolidated` method, to show you what I mean:

```
/// Do a "consolidated" STX transfer.
/// If the 'from' principal has locked STX, and they have unlocked, then process the STX unlock
/// and update its balance in addition to spending tokens out of it.
pub fn stx_transfer_consolidated(env: &mut Environment, from: &PrincipalData, to: &PrincipalData, amount: u128) -> Result<Value> {
    if amount <= 0 {
        return clarity_ecode!(StxErrorCodes::NON_POSITIVE_AMOUNT);
    }

    if from == to {
        return clarity_ecode!(StxErrorCodes::SENDER_IS_RECIPIENT);
    }

    if Some(from.clone()) != env.sender.as_ref().map(|pval| pval.clone().expect_principal()) {
        return clarity_ecode!(StxErrorCodes::SENDER_IS_NOT_TX_SENDER);
    }

    let (from_bal, unlock) = stx_balance_consolidated(&mut env.global_context.database, from);
    let to_bal = env.global_context.database.get_account_stx_balance(to);

    if from_bal < amount {
        return clarity_ecode!(StxErrorCodes::NOT_ENOUGH_BALANCE)
    }

    let final_from_bal = from_bal - amount;
    let final_to_bal = to_bal.checked_add(amount)
        .ok_or(RuntimeErrorType::ArithmeticOverflow)?;

    // loading from/to principals and balances
    env.add_memory(TypeSignature::PrincipalType.size() as u64)?;
    env.add_memory(TypeSignature::PrincipalType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;

    // loading from's locked amount and height
    // TODO: this does not count the inner stacks block header load, but arguably,
    // this could be optimized away, so it shouldn't penalize the caller.
    env.add_memory(TypeSignature::UIntType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;

    // NOTE: this updates the balance with the unlocked tokens, if we did an unlock.
    env.global_context.database.set_account_stx_balance(from, final_from_bal);
    env.global_context.database.set_account_stx_balance(to,   final_to_bal);

    if unlock {
        // clear lock state
        env.global_context.database.set_account_stx_locked(from, 0);
        env.global_context.database.set_account_unlock_height(from, 0);
        debug!("Consolidated {} after stx-transfer", from);
    }

    env.global_context.log_stx_transfer(&from, amount)?;
    env.register_stx_transfer_event(from.clone(), to.clone(), amount)?;
    Ok(Value::okay_true())
}
```

Notice that only the _sender's_ lock state is loaded and stored.  The only data loaded and stored for the recipient (`to`) is the (unlocked) token balance.  That was intentional.

Per https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r479542332, it would be fine to consolidate both the STX locked and block height into a single MARF key/value pair, since they are always loaded/stored together anyway.  But, I do not think the three fields should be combined, since that would penalize the sender, who would then be forced to at least load the lock state for a recipient _even though it would not be used_.   I'll update the comment to reflect this as well.

Even if we did consolidate all three fields and (IMHO unnecessarily) penalize the sender for sending to a recipient, in no case should the sender be paying to unlock the recipient's tokens.  Storing data is even more costly than loading it.

Reply:
I should have phrase differently :). I noticed and **understood** that in your first implementation, you **did not** materialize the lock update, because it **would** have cost to the sender, and this would have been unfair.
Addressing your initial question: I took a different approach in this implementation, and did chose to update the lock because the balance's recipient (including unlocked_amount, locked_amount and unlock_height) need to be get and set, since we're updating the unlocked_amount.

Reply:
I'm glad we're on the same page :) 

> Addressing your initial question: I took a different approach in this implementation, and did chose to update the lock because the balance's recipient (including unlocked_amount, locked_amount and unlock_height) need to be get and set, since we're updating the unlocked_amount.

I consider this a design flaw, then.  The sender should not be paying to unlock the recipient's tokens.  Doing so would increase the runtime read cost of a `TokenTransfer` by 33%, and write cost by up to 33%

Before: load 4 u128's (sender balance, locked tokens, unlock block height; recipient balance), store either 2 or 4 u128s (sender and recipient balance, _possibly also_ sender locked tokens and sender unlock block height if the sender's tokens unlocked).

After:  load 6 u128's (sender balance, locked tokens, unlock block height; recipient balance, locked tokens, unlock block height); store either 2, 4, or 6 u128's (sender and recipient balance, _possibly also_ sender locked tokens and sender unlock block height if the sender's tokens unlocked, and _possibly also_ the recipient locked tokens and recipient unlock block height if the recipient's tokens unlocked).

It may not seem like much, but in aggregate it raises transaction fees for everyone.  All of these I/O operations come out of the block's total compute budget.

Reply:
I feel like there's a tradeoff to be made, and we should look at the big picture. We have 2 main use cases:

STX-TRANSFER (require a transaction fee)
- Before: read 4 MARF entries of 16 bytes, write 4 MARF entries of 16 bytes
- After: read 2 MARF entries of 48 bytes, write 2 MARF entries of 48 bytes

STX-GET-BALANCE (free of charge)
- Before: read 3 MARF entries of 16 bytes
- After: read 1 MARF entry of 48 bytes

We should consider the fact that the ratio of calls `stx-transfer`:`stx-get-balance` is far from 1:1 (1:10? 1:100?).

This feedback made sense when I started addressing it, because I was **under the impression** that MARF reading 3 entries of 16 bytes was more expensive than MARF reading 1 entry of 48 bytes.

Reply:
I agree with @lgalabru here -- the additional 32 bytes read and serialized out of the database are _very_ small compared to the cost of additional MARF reads or writes.

Reply:
I think something got lost here.  I'm talking about the costs that the VM tracks for calculating how full a block is.  The fundamental issue this comment thread is about is making sure that we keep the VM-accounted runtime costs as low as possible for the transaction sender.  Unless I'm missing something, this change is making it worse.

@kantai are you saying that 1 MARF read for 32 bytes is tabulated by the VM as a cheaper operation than 3 MARF reads for 16 bytes?

Reply:
Resolution: we will weight MARF read/write operations more heavily than byte read/write operations, since they are considerably more expensive.  This would mean that 1 MARF read for 48 bytes would be considerably cheaper to the VM than 3 MARF reads for 16 bytes.

Reply:
From the public blockchain meeting, we'll keep the current STX balance representation, and we'll be sure to weight MARF I/O operations much more heavily (i.e. consider them much more expensive than) raw I/O on the sidestore (https://github.com/blockstack/stacks-blockchain/issues/1877).